### PR TITLE
Add unit to course

### PR DESCRIPTION
**Pull request by tjingsheng**

### CODE REVIEW

File: src/main/java/seedu/address/model/localcourse/LocalCourse.java [hunk 1]
```diff
@@ -73,7 +79,8 @@ public boolean equals(Object other) {
 
         LocalCourse otherLocalCourse = (LocalCourse) other;
         return localCode.equals(otherLocalCourse.localCode)
-                && localName.equals(otherLocalCourse.localName);
+                && localName.equals(otherLocalCourse.localName)
+                && localUnit.equals(otherLocalCourse.localUnit);
     }
 
     @Override
```
Comment (dlathyun) [inline src/main/java/seedu/address/model/localcourse/LocalCourse.java:86]: Is there any specific reason why LocalUnit is not included in this method?

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java [hunk 2]
```diff
@@ -89,7 +92,8 @@ public void parse_invalidValue_failure() {
         // invalid localCode
         assertParseFailure(parser, commandActionWord
                         + getSquareBracketWrappedArgument(INVALID_LOCAL_COURSE_CODE)
-                        + " " + getSquareBracketWrappedArgument(TYPICAL_LOCAL_COURSE_NAME),
+                        + " " + getSquareBracketWrappedArgument(TYPICAL_LOCAL_COURSE_NAME)
+                        + getSquareBracketWrappedArgument(TYPICAL_LOCAL_COURSE_UNIT.toString()),
```
Comment (dlathyun) [inline src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:96]: Should there be space " " in between?

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java [hunk 3]
```diff
@@ -109,7 +113,8 @@ public void parse_invalidValue_failure() {
     public void parse_edgeCaseLocalNameValue_success() {
         assertParseSuccess(parser, commandActionWord
                         + getSquareBracketWrappedArgument(EDGE_CASE_VALID_LOCAL_COURSE_CODE)
-                        + " " + getSquareBracketWrappedArgument(EDGE_CASE_VALID_LOCAL_COURSE_NAME),
+                        + " " + getSquareBracketWrappedArgument(EDGE_CASE_VALID_LOCAL_COURSE_NAME)
+                        + getSquareBracketWrappedArgument(EDGE_CASE_VALID_LOCAL_COURSE_UNIT.toString()),
```
Comment (dlathyun) [inline src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:117]: here also!

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java [hunk 4]
```diff
@@ -43,7 +45,8 @@ public void parse_unnecessaryWhiteSpace_success() {
                         + getSquareBracketWrappedArgument(expectedLocalCourse.getLocalCode().getValue())
                         + UNNCESSARY_WHITESPACE
                         + getSquareBracketWrappedArgument(expectedLocalCourse.getLocalName().getValue())
-                        + UNNCESSARY_WHITESPACE,
+                        + UNNCESSARY_WHITESPACE
+                        + getSquareBracketWrappedArgument(expectedLocalCourse.getLocalUnit().getValue().toString()),
```
Comment (dlathyun) [inline src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:49]: How about putting getLocalUnit().toString() straightaway?

File: src/test/java/seedu/address/storage/JsonAdaptedLocalCourseTest.java [hunk 5]
```diff
@@ -26,31 +27,52 @@ public void toModelType_validLocalCourseDetails_returnsLocalCourse() throws Exce
...
+    @Test
+    public void toModelType_negativeLocalUnit_throwsIllegalValueException() {
+        JsonAdaptedLocalCourse localCourse = new JsonAdaptedLocalCourse(
+            TYPICAL_LOCAL_COURSE_CODE,
+            null,
+            TYPICAL_LOCAL_COURSE_UNIT);
+        String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, LocalName.class.getSimpleName());
+        assertThrows(IllegalValueException.class, expectedMessage, localCourse::toModelType);
+    }
 }
```
Comment (dlathyun) [inline src/test/java/seedu/address/storage/JsonAdaptedLocalCourseTest.java:78]: Great testing on the negative LocalUnit! Maybe can add tests for null value and non-string/number values also?

File: src/test/java/seedu/address/storage/JsonAdaptedPartnerCourseTest.java [hunk 6]
```diff
@@ -27,33 +28,46 @@ public void toModelType_validPartnerCourseDetails_returnsPartnerCourse() throws
...
     }
 
     @Test
     public void toModelType_nullPartnerName_throwsIllegalValueException() {
         JsonAdaptedPartnerCourse partnerCourse = new JsonAdaptedPartnerCourse(
-                TYPICAL_UNIVERSITY_NAME, TYPICAL_PARTNER_COURSE_CODE, null);
+            TYPICAL_UNIVERSITY_NAME,
+            TYPICAL_PARTNER_COURSE_CODE,
+            null,
+            TYPICAL_PARTNER_COURSE_UNIT);
```
Comment (dlathyun) [inline src/test/java/seedu/address/storage/JsonAdaptedPartnerCourseTest.java:70]: Maybe can add more PartnerUnit related tests!

File: src/test/java/seedu/address/testutil/PartnerCourseUtil.java [hunk 7]
```diff
@@ -12,40 +12,35 @@ public class PartnerCourseUtil {
...
      * Returns the part of add command string for the given {@code PartnerCourse}'s details.
      */
     public static String getPartnerCourseArgumentsForAddCommand(PartnerCourse partnerCourse) {
-        StringBuilder sb = new StringBuilder("[");
-        sb.append(partnerCourse.getPartnerCode()).append("] [");
-        sb.append(partnerCourse.getPartnerName()).append("]");
-        return sb.toString();
+        return String.format("[%s] [%s]",
+            partnerCourse.getPartnerCode(),
+            partnerCourse.getPartnerName());
```
Comment (dlathyun) [inline src/test/java/seedu/address/testutil/PartnerCourseUtil.java:27]: I think there should be PartnerUnit as well!

Comment (dlathyun) [reply src/test/java/seedu/address/testutil/PartnerCourseUtil.java:27]: wait and university haha.. my bad for not putting that earlier

File: src/main/java/seedu/address/model/partnercourse/PartnerUnit.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            return false;
+        }
+
+        if (test instanceof String) {
+            // If the input is a String, attempt to parse it as a double
+            try {
+                double parsedValue = Double.parseDouble((String) test);
+
+                // Check if the parsed value is greater than 0
+                return parsedValue >= 0;
```
Comment (dlathyun) [inline src/main/java/seedu/address/model/partnercourse/PartnerUnit.java:60]: Actually may I ask why we are accepting 0.0 as valid Local/PartnerUnit?

**Review by dlathyun — CHANGES_REQUESTED** [review]
Well done in adding the extra Local/PartnerUnit! 
One thing to take note is that I think we might have to update the json files in data folder too

File: src/test/java/seedu/address/logic/parser/PartnerCourseAddCommandParserTest.java [hunk 9]
```diff
@@ -115,7 +119,8 @@ public void parse_edgeCasePartnerNameValue_success() {
         assertParseSuccess(parser, commandActionWord
                         + getSquareBracketWrappedArgument(EDGE_CASE_VALID_UNIVERSITY_NAME) + " "
                         + getSquareBracketWrappedArgument(EDGE_CASE_VALID_PARTNER_COURSE_CODE)
-                        + " " + getSquareBracketWrappedArgument(EDGE_CASE_VALID_PARTNER_COURSE_NAME),
+                        + " " + getSquareBracketWrappedArgument(EDGE_CASE_VALID_PARTNER_COURSE_NAME)
+                        + getSquareBracketWrappedArgument(EDGE_CASE_VALID_PARTNER_COURSE_UNIT.toString()),
```
[the remark this answers is not in the document]
Comment (tjingsheng) [reply src/test/java/seedu/address/logic/parser/PartnerCourseAddCommandParserTest.java:123]: Good catch, I think even though it does not affect the result of test case, it is not neccessary, I shall remove them from the partnercourse add and localcourse add tests

File: src/main/java/seedu/address/logic/Messages.java [hunk 10]
```diff
@@ -81,12 +85,15 @@ public static String format(University university) {
...
-                .append(partnerCourse.getPartnerName());
+        final StringBuilder builder = new StringBuilder();
+        builder.append("PartnerUniversity: ")
+            .append(partnerCourse.getPartnerUniversity().getUniversityName().toString())
+            .append("; PartnerCode: ")
+            .append(partnerCourse.getPartnerCode())
+            .append("; PartnerName: ")
+            .append(partnerCourse.getPartnerName())
+            .append("; LocalUnit: ")
+            .append(partnerCourse.getPartnerUnit());
```
[the remark this answers is not in the document]
Comment (tjingsheng) [reply src/main/java/seedu/address/logic/Messages.java:96]: yes, good catch

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 11]
```diff
@@ -190,6 +207,21 @@ public static PartnerName parsePartnerName(String partnerName) throws ParseExcep
         return new PartnerName(partnerName);
     }
 
+    /**
+     * Parses a {@code String partnerUnit} into an {@code PartnerUnit}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code localName} is invalid.
```
[the remark this answers is not in the document]
Comment (tjingsheng) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:214]: yes

File: src/main/java/seedu/address/model/localcourse/LocalCourse.java (hunk 1 above)
Comment (tjingsheng) [reply src/main/java/seedu/address/model/localcourse/LocalCourse.java:86]: localunit has been included, not too sure which you are refering too..

File: src/main/java/seedu/address/model/partnercourse/PartnerCourse.java [hunk 12]
```diff
@@ -71,7 +80,8 @@ public boolean equals(Object other) {
         PartnerCourse otherPartnerCourse = (PartnerCourse) other;
         return partnerUniversity.equals(otherPartnerCourse.partnerUniversity)
                 && partnerCode.equals(otherPartnerCourse.partnerCode)
-                && partnerName.equals(otherPartnerCourse.partnerName);
+                && partnerName.equals(otherPartnerCourse.partnerName)
+                && partnerUnit.equals(otherPartnerCourse.partnerUnit);
     }
 
     @Override
```
[the remark this answers is not in the document]
Comment (tjingsheng) [reply src/main/java/seedu/address/model/partnercourse/PartnerCourse.java:87]: as above

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java (hunk 2 above)
Comment (tjingsheng) [reply src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:96]: as above

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java (hunk 3 above)
Comment (tjingsheng) [reply src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:117]: same as above

File: src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java (hunk 4 above)
Comment (tjingsheng) [reply src/test/java/seedu/address/logic/parser/LocalCourseAddCommandParserTest.java:49]: good catch, in fact, I should be using the toString() method for the other fields as well instead of using getValue

File: src/test/java/seedu/address/testutil/PartnerCourseUtil.java (hunk 7 above)
Comment (tjingsheng) [reply src/test/java/seedu/address/testutil/PartnerCourseUtil.java:27]: done

File: src/main/java/seedu/address/model/partnercourse/PartnerUnit.java (hunk 8 above)
Comment (tjingsheng) [reply src/main/java/seedu/address/model/partnercourse/PartnerUnit.java:60]: 0 mc mods exist :P

**Review by dlathyun — APPROVED** [review]
LGTM! Nice changes with getValue()/toString() methods.
